File: php/src/Providers/TempmailPlusProvider.php
<?php

declare(strict_types=1);

namespace TempMail\Providers;

use TempMail\HttpClient;
use TempMail\Models\Message;
use TempMail\Models\MessageDetail;
use TempMail\TempMailProvider;
use TempMail\Exceptions\NotFoundException;
use TempMail\Exceptions\TempMailException;

class TempmailPlusProvider implements TempMailProvider
{
    private const BASE_URL = 'https://tempmail.plus';
    private const DOMAIN = 'mailto.plus';

    private HttpClient $http;
    private ?string $currentEmail = null;

    public function __construct(?HttpClient $http = null)
    {
        $this->http = $http ?? new HttpClient();
    }

    public function generateEmail(): string
    {
        $username = bin2hex(random_bytes(5)); // 10 chars
        $this->currentEmail = "{$username}@" . self::DOMAIN;
        return $this->currentEmail;
    }

    public function getInbox(string $email): array
    {
        $res = $this->http->get(self::BASE_URL . '/api/mails?' . http_build_query(['email' => $email]));
        $data = $res['body'];
        if (is_array($data) && ($data['result'] ?? true) === false) {
            throw new TempMailException('tempmail.plus: API returned error');
        }
        $messages = [];
        foreach ($data['mail_list'] ?? [] as $item) {
            $messages[] = new Message(
                id: (string)($item['mail_id'] ?? ''),
                sender: $item['from_mail'] ?? '',
                subject: $item['subject'] ?? '',
                date: $this->parseDate($item['time'] ?? null),
            );
        }
        return $messages;
    }

    public function readMessage(string $email, string $messageId): MessageDetail
    {
        $res = $this->http->get(self::BASE_URL . "/api/mails/{$messageId}?" . http_build_query(['email' => $email]));
        $data = $res['body'];
        if (!is_array($data)) {
            throw new NotFoundException("tempmail.plus: message {$messageId} not found");
        }
        $attachments = [];
        foreach ($data['attachments'] ?? [] as $a) {
            $attachments[] = [
                'name' => $a['filename'] ?? '',
                'type' => $a['content_type'] ?? '',
                'size' => $a['size'] ?? 0,
            ];
        }
        return new MessageDetail(
            id: (string)($data['mail_id'] ?? $messageId),
            sender: $data['from_mail'] ?? $data['from'] ?? '',
            subject: $data['subject'] ?? '',
            date: $this->parseDate($data['date'] ?? null),
            bodyText: $data['text'] ?? '',
            bodyHtml: $data['html'] ?? '',
            attachments: $attachments,
        );
    }

    public function deleteEmail(string $email): bool
    {
        $this->currentEmail = null;
        return true;
    }

    public function waitForEmail(string $email, int $timeout = 60, int $interval = 5): ?Message
    {
        $deadline = time() + $timeout;
        while (time() < $deadline) {
            if (!empty($inbox = $this->getInbox($email))) {
                return $inbox[0];
            }
            sleep($interval);
        }
        return null;
    }

    private function parseDate(?string $str): \DateTimeImmutable
    {
        if ($str !== null && $str !== '') {
            try {
                return new \DateTimeImmutable($str);
            } catch (\Throwable) {}
        }
        return new \DateTimeImmutable();
    }
}

File: java/src/main/java/com/tempmail/providers/TempmailPlusProvider.java
package com.tempmail.providers;

import com.tempmail.TempMailProvider;
import com.tempmail.errors.NotFoundException;
import com.tempmail.errors.TempMailException;
import com.tempmail.http.HttpClient;
import com.tempmail.json.SimpleJson;
import com.tempmail.json.SimpleJson.JsonArray;
import com.tempmail.json.SimpleJson.JsonObject;
import com.tempmail.json.SimpleJson.JsonValue;
import com.tempmail.models.Message;
import com.tempmail.models.MessageDetail;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Random;

/**
 * tempmail.plus provider - REST API, no auth, email as query param.
 */
public class TempmailPlusProvider implements TempMailProvider {

    private static final String BASE_URL = "https://tempmail.plus";
    private static final String DOMAIN = "mailto.plus";
    private static final String CHARS = "abcdefghijklmnopqrstuvwxyz0123456789";
    private static final DateTimeFormatter DATE_FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    private final HttpClient client;
    private final Random random = new Random();
    private String email;

    public TempmailPlusProvider() {
        this.client = new HttpClient(true, true);
    }

    @Override
    public String generateEmail() throws TempMailException {
        StringBuilder sb = new StringBuilder(10);
        for (int i = 0; i < 10; i++) sb.append(CHARS.charAt(random.nextInt(CHARS.length())));
        email = sb.toString() + "@" + DOMAIN;
        return email;
    }

    @Override
    public List<Message> getInbox(String email) throws TempMailException {
        JsonObject data = SimpleJson.parseObject(client.get(BASE_URL + "/api/mails?email=" + enc(email)));
        if (!data.getBoolean("result")) {
            // ponytail: result defaults to true on success; only flag errors when explicitly false
            if (data.has("result")) throw new TempMailException("tempmail.plus: API returned error");
        }
        JsonArray items = data.getArray("mail_list");
        List<Message> messages = new ArrayList<>();
        for (JsonValue el : items.elements()) {
            JsonObject item = el.asObject();
            messages.add(new Message(
                    item.getString("mail_id"),
                    item.getString("from_mail"),
                    item.getString("subject"),
                    parseDate(item.getString("time"))
            ));
        }
        return messages;
    }

    @Override
    public MessageDetail readMessage(String messageId) throws TempMailException {
        if (email == null) throw new TempMailException("tempmail.plus: no email - call generateEmail() first");
        JsonObject data = SimpleJson.parseObject(client.get(
                BASE_URL + "/api/mails/" + enc(messageId) + "?email=" + enc(email)));
        String text = data.getString("text");
        String html = data.getString("html");
        return new MessageDetail(
                data.has("mail_id") ? data.getString("mail_id") : messageId,
                str(data.get("from_mail"), data.get("from")),
                data.getString("subject"),
                parseDate(str(data.get("time"), data.get("date"))),
                text != null ? text : "",
                html != null ? html : "",
                extractAttachments(data)
        );
    }

    @Override
    public boolean deleteEmail(String email) throws TempMailException {
        this.email = null;
        return true;
    }

    @Override
    public Optional<Message> waitForEmail(String email, Duration timeout, Duration interval) throws TempMailException {
        long deadline = System.currentTimeMillis() + timeout.toMillis();
        while (System.currentTimeMillis() < deadline) {
            List<Message> inbox = getInbox(email);
            if (!inbox.isEmpty()) return Optional.of(inbox.get(0));
            try {
                Thread.sleep(interval.toMillis());
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new TempMailException("Interrupted while waiting for email", e);
            }
        }
        return Optional.empty();
    }

    private List<Map<String, Object>> extractAttachments(JsonObject data) {
        JsonArray arr = data.getArray("attachments");
        List<Map<String, Object>> out = new ArrayList<>();
        for (JsonValue el : arr.elements()) {
            JsonObject a = el.asObject();
            Map<String, Object> m = new HashMap<>();
            m.put("filename", a.getString("filename"));
            m.put("content_type", a.getString("content_type"));
            m.put("size", a.getInt("size"));
            out.add(m);
        }
        return out;
    }

    private String enc(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }

    private String str(JsonValue... vals) {
        for (JsonValue v : vals) {
            if (v != null && !v.isNull()) {
                String s = v.asString();
                if (s != null && !s.isEmpty()) return s;
            }
        }
        return "";
    }

    private LocalDateTime parseDate(String s) {
        if (s == null || s.isEmpty()) return LocalDateTime.now();
        try {
            return LocalDateTime.parse(s, DATE_FMT);
        } catch (Exception ignore) {}
        try {
            return LocalDateTime.parse(s.replace("Z", "").replace("+00:00", ""));
        } catch (Exception ignore) {}
        try {
            return LocalDateTime.parse(s, DateTimeFormatter.RFC_1123_DATE_TIME);
        } catch (Exception ignore) {}
        return LocalDateTime.now();
    }
}
